core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/result/ItemizedFileDocument.scala

```diff
@@ -0,0 +1,202 @@
+package edu.uci.ics.amber.core.storage.result
+
+import com.twitter.chill.{KryoPool, ScalaKryoInstantiator}
+import edu.uci.ics.amber.core.storage.model.{BufferedItemWriter, VirtualDocument}
+import org.apache.commons.vfs2.{FileObject, VFS}
+
+import java.io.{DataOutputStream, InputStream}
+import java.net.URI
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import scala.collection.mutable.ArrayBuffer
+
+object ItemizedFileDocument {
+  // Initialize KryoPool as a static object
+  private val kryoPool = KryoPool.withByteArrayOutputStream(10, new ScalaKryoInstantiator)
+}
+
+/**
+  * ItemizedFileDocument provides methods to read/write items to a file located on the filesystem.
+  * All methods are THREAD-SAFE, implemented using a read-write lock:
+  * - 1 writer at a time: only 1 thread can acquire the write lock.
+  * - n readers at a time: multiple threads can acquire the read lock.
+  *
+  * The type parameter T specifies the iterable data item stored in the file.
+  *
+  * @param uri the identifier of the file. If the file doesn't physically exist, ItemizedFileDocument will create it during construction.
+  */
+class ItemizedFileDocument[T >: Null <: AnyRef](val uri: URI)
+    extends VirtualDocument[T]
+    with BufferedItemWriter[T] {
+
+  val file: FileObject = VFS.getManager.resolveFile(uri.toString)
+  val lock = new ReentrantReadWriteLock()
+
+  // Buffer to store items before flushing
+  private val buffer = new ArrayBuffer[T]()
+  override val bufferSize: Int = 1024
+
+  // Register a shutdown hook to delete the file when the JVM exits
+  sys.addShutdownHook {
```

Comment:
The lifecycle for this file is also not correct. This file is created by the computing unit JVM, which can be killed right after the execution.

Reply:
I suggest we do a global cleanup on `OpResultStorage` level based on #3146.

Reply:
This file is removed.